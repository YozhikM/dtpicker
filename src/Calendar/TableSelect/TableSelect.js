/* @flow */

import React from 'react';
import './TableSelect.scss';

export type Value = number;
export type Options = Array<{ value: Value, name: string }>;

type Props = {|
  cols: number,
  options: Options,
  value: Value,
  onChange?: (value: Value) => void,
|};
type State = {|
  value: Value,
|};

class TableSelect extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      value: this.props.value,
    };
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.value !== this.state.value) {
      this.setState({ value: nextProps.value });
    }
  }

  clickItem(e: Event, value: Value) {
    this.setState({ value });
    const { onChange } = this.props;
    if (onChange) {
      onChange(value);
    }
  }

  render() {
    const { cols, options } = this.props;
    const { value } = this.state;

    return (
      <div className="root">
        <ul
          style={{
            width: `${cols * 50}px`,
          }}
        >
          {options.map((option, i) => (
            <li
              onClick={e => {
                this.clickItem(e, option.value);
              }}
              key={i}
              style={{
                fontWeight: option.value === value ? 'bold' : 'normal',
              }}
            >
              {option.name}
            </li>
          ))}
        </ul>
      </div>
    );
  }
}

export default TableSelect;
