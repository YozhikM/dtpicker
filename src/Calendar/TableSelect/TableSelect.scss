.root {
  width: 200px;

  ul {
    list-style: none;
    display: flex;
    flex-flow: row wrap;
    align-items: center;
    justify-content: center;
    margin: 0 auto;
    padding: 0;
  }

  li {
    display: flex;
    justify-content: center;
    align-items: center;
    text-align: center;
    width: 50px;
    height: 50px;
    cursor: pointer;
    border: 1px solid rgb(228, 231, 231);
    box-sizing: border-box;
  }

  li:hover {
    background-color: lighten(#34495e, 30%);
    color: #fff;
  }
}
