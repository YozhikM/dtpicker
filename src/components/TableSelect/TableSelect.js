/* @flow */

import React from 'react';
import s from './TableSelect.scss';

export type Value = number;
export type Options = Array<{ value: Value, name: string }>;

type Props = {|
  cols: number,
  options: Options,
  value: Value,
  onChange?: (value: Value) => void
|};
type State = {|
  value: Value
|};

class TableSelect extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      value: this.props.value
    };
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.value !== this.state.value) {
      this.setState({ value: nextProps.value });
    }
  }

  clickItem(e: Event, value: Value) {
    this.setState({ value });
    const { onChange } = this.props;
    if (onChange) {
      onChange(value);
    }
  }

  render() {
    const { cols, options } = this.props;
    const { value } = this.state;

    return (
      <div className={s.root}>
        <ul
          style={{
            width: `${cols * 68.25}px`
          }}
        >
          {options.map((options, i) => {
            return (
              <li
                onClick={e => {
                  this.clickItem(e, options.value);
                }}
                key={i}
                style={{
                  fontWeight: options.value === value ? 'bold' : 'normal'
                }}
              >
                {options.name}
              </li>
            );
          })}
        </ul>
      </div>
    );
  }
}

export default TableSelect;
