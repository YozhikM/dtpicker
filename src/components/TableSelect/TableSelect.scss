.root {
    display: inline-block;
    width: 50%;
  ul {
    list-style: none;
    display: flex;
    flex-flow: row wrap;
    align-items: center;
    margin: 0;
    padding: 0;
  }

  li {
    display: flex;
    justify-content: center;
    align-items: center;
    width: 75px;
    height: 75px;
    cursor: pointer;
    font-size: .75rem;
  }

  li:hover {
    background-color: #333;
    color: #fff;
  }
}
