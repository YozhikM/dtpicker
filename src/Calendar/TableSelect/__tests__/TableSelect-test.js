/* @flow */

import React from 'react';
import TableSelect from '../TableSelect';

describe('TableSelect', () => {
  describe('render()', () => {
    it('basic', () => {
      const options = [{ value: 99, name: 'ninety nine' }];

      expect(shallow(<TableSelect cols={1} options={options} value={2} />)).toMatchSnapshot();
    });

    it('if matched', () => {
      const options = [{ value: 2, name: 'two' }];

      expect(shallow(<TableSelect cols={1} options={options} value={2} />)).toMatchSnapshot();
    });

    it('if 2 cols', () => {
      const options = [{ value: 21, name: 'twenty one' }];

      expect(shallow(<TableSelect cols={2} options={options} value={2} />)).toMatchSnapshot();
    });
  });

  it('componentWillReceiveProps()', () => {
    const options = [{ value: 1, name: 'one' }];
    const wrapper = shallow(<TableSelect cols={1} options={options} value={1} />);
    const nextProps = {
      cols: 1,
      value: 3,
      options,
    };
    wrapper.instance().componentWillReceiveProps(nextProps);

    expect(wrapper.state('value')).toBe(3);
  });

  describe('user interactions', () => {
    it('clickItem()', () => {
      const options = [{ value: 10, name: 'ten' }];
      const onChange = jest.fn();
      const wrapper = shallow(
        <TableSelect cols={1} options={options} onChange={onChange} value={2} />
      );
      wrapper.find('li').simulate('click');

      expect(wrapper.state('value')).toBe(10);
      expect(onChange).toBeCalledWith(10);
    });
  });
});
