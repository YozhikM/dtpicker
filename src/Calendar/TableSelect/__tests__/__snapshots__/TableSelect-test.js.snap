// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableSelect render() basic 1`] = `
<div>
  <ul
    style={
      Object {
        "width": "50px",
      }
    }
  >
    <li
      key="ninety nine"
      onClick={[Function]}
      style={
        Object {
          "fontWeight": "normal",
        }
      }
    >
      ninety nine
    </li>
  </ul>
</div>
`;

exports[`TableSelect render() if 2 cols 1`] = `
<div>
  <ul
    style={
      Object {
        "width": "100px",
      }
    }
  >
    <li
      key="twenty one"
      onClick={[Function]}
      style={
        Object {
          "fontWeight": "normal",
        }
      }
    >
      twenty one
    </li>
  </ul>
</div>
`;

exports[`TableSelect render() if matched 1`] = `
<div>
  <ul
    style={
      Object {
        "width": "50px",
      }
    }
  >
    <li
      key="two"
      onClick={[Function]}
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      two
    </li>
  </ul>
</div>
`;
